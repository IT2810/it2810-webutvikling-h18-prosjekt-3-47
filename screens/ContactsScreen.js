
import React from 'react';
import {
    Platform,
    StyleSheet,
    Text,
    TouchableOpacity,
    View,
    Button,
    AsyncStorage
} from 'react-native';

import PopupDialog from 'react-native-popup-dialog';
import DialogTitle from "react-native-popup-dialog/src/components/DialogTitle";

import ContactList from '../components/ContactList';

import CreateContactScreen from './CreateContactScreen';


export default class ContactsScreen extends React.Component {
    static navigationOptions = {
        title: 'Contacts',
    };

    constructor(props) {
        super(props);

        this.state = {
            contacts: [],
            popupDialog: null
        };

        // Calling this inside receiveNewContact will point to ContactScreen
        this.receiveNewContact = this.receiveNewContact.bind(this);
    }

    /**
     * Loading existing contacts using AsyncStorage, into the state of ContactScreen
     *
     */
    _loadContacts = async () => {
        try {
            const value = await AsyncStorage.getItem('contacts');
            if (value !== null) {
                //console.log(value);
                //console.log(JSON.parse(value));
                this.setState(() => ({
                    contacts: JSON.parse(value)
                }));
            }
        } catch (error) {
            console.log('AsyncStorage error while loading: ' + error.message);
        }
    };

    // Placeholders
    contacts = [
        {
            name: "Navn Navnesen 1",
            phoneNumber: 12345678,
            address: "Adresse 1"
        },
        {
            name: "Navn Navnesen 2",
            phoneNumber: 87654321,
            address: "Adresse 2"
        }
    ];

    /**
     *  Storing the current contacts using AsyncStorage, from the state of ContactScreen
     *
     */
    _storeContact = async () => {
        try {
            await AsyncStorage.setItem('contacts', JSON.stringify(this.contacts));
        } catch (error) {
            console.log('AsyncStorage error while storing: ' + error.message);
        }
    };

    /**
     *  Receives a newly created contact object from CreateContactScreen, and push it to the state of ContactScreen
     *  @param contact <object>
     */
    receiveNewContact(contact) {
        let contacts = this.state.contacts.slice();
        contacts.push(contact);
        this.setState({contacts: contacts});
    }

    componentDidMount() {
        //this._storeContact();
        this._loadContacts();
    }

    render() {
        return (
            <View style={styles.container}>
                <ContactList contacts={this.state.contacts}/>
                <Button
                    title='Create New Contact'
                    onPress={() => {
                        this.state.popupDialog.show();
                    }}
                />
                <PopupDialog
                    dialogTitle={<DialogTitle title="New Contact"/>}
                    ref={(popupDialog) => {
                        if (!this.state.popupDialog){
                            this.setState({
                                popupDialog: popupDialog
                            })
                        }
                    }}
                    height={Number(400)}>
                    <CreateContactScreen callback={this.receiveNewContact} popupDialog={this.state.popupDialog}/>
                </PopupDialog>
            </View>
        );
    }
}

const styles = StyleSheet.create({
  container: {
      flex: 1,
      backgroundColor: '#fff',
      justifyContent: 'center',
      alignItems: 'center',
      paddingBottom: 10,
  },
})
