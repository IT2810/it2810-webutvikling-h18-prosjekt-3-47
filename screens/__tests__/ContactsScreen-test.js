import 'react-native';
import React from 'react';
import renderer from 'react-test-renderer';
import ShallowRenderer from 'react-test-renderer/shallow';
import ContactsScreen from '../ContactsScreen';

jest.mock("react-native-popup-dialog");

let shallowRenderer = new ShallowRenderer();

describe('Rendering', () => {
    it('renders correctly', () => {
        const tree = renderer.create(<ContactsScreen />).toJSON();

        expect(tree).toMatchSnapshot();
    });

    it('renders shallow correctly', () => {
        shallowRenderer.render(<ContactsScreen />);
        const result = shallowRenderer.getRenderOutput();

        expect(result).toMatchSnapshot();
    });
});


describe('Logical', () => {
    it('should have an empty contact list on first render', () => {

        let contactsScreen = renderer.create(<ContactsScreen/>);
        let contactsScreenInstance = contactsScreen.getInstance();

        expect(contactsScreenInstance.state.contacts).toEqual([]);
    });
});
