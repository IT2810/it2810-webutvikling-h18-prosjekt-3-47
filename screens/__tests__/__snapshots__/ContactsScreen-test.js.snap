// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Rendering renders correctly 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#fff",
      "flex": 1,
      "justifyContent": "center",
      "paddingBottom": "3%",
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "#fff",
        "flex": 1,
        "flexDirection": "row",
        "padding": "3%",
      }
    }
  />
  <View
    accessibilityComponentType="button"
    accessibilityTraits={
      Array [
        "button",
      ]
    }
    accessible={true}
    isTVSelectable={true}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "opacity": 1,
      }
    }
  >
    <View
      style={
        Array [
          Object {},
        ]
      }
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Array [
            Object {
              "color": "#007AFF",
              "fontSize": 18,
              "padding": 8,
              "textAlign": "center",
            },
          ]
        }
      >
        Create New Contact
      </Text>
    </View>
  </View>
</View>
`;

exports[`Rendering renders shallow correctly 1`] = `
<Component
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#fff",
      "flex": 1,
      "justifyContent": "center",
      "paddingBottom": "3%",
    }
  }
>
  <ContactList
    contacts={Array []}
  />
  <Button
    onPress={[Function]}
    title="Create New Contact"
  />
  <PopupDialog
    dialogTitle={
      <DialogTitle
        hasTitleBar={true}
        title="New Contact"
        titleAlign="center"
      />
    }
    height={400}
  >
    <CreateContact
      callback={[Function]}
      popupDialog={null}
    />
  </PopupDialog>
</Component>
`;
